### ISSUE TITLE
What is the best way to bridge complex structs into (and out of) goja?

### ISSUE THREAD

**Issue by benallfree** [dop251/goja#434]
This code is from [PocketBase](https://pocketbase.io/docs/event-hooks/#app-hooks):

```go
    app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
        // add new "GET /api/hello" route
		
        e.Router.AddRoute(echo.Route{
            Method: http.MethodGet,
            Path:   "/api/go",
            Handler: func(c echo.Context) error {
                return c.String(200, "Hello from go!")
            },
            Middlewares: []echo.MiddlewareFunc{
                apis.RequireAdminOrUserAuth(),
            },
        })

        return nil
    })
```

My first attempt was just this:

```go
vm.Set("app", app)
vm.Set("http", http)
vm.Set("apis", apis)
```

```ts
app.OnBeforeServe().Add(e=>{
   e.Router.AddRoute({
    Method: http.MethodGet,
    Path: `/api/js`,
    Handler: c=>{
      c.String(200, `Hello from JS`
    },
    Middlewares: [apis.RequireAdminOrUserAuth()]
})
```

But that failed on `AddRoute` because it couldn't convert the JS object back into an `echo.Route` struct. So then I added a constructor function:

```go
	vm.Set("newRoute", func(Method string, Path string, Handler echo.HandlerFunc, Middlewares []echo.MiddlewareFunc ) echo.Route {
		return echo.Route{
            Method: Method,
            Path:  Path,
            Handler: Handler,
            Middlewares: Middlewares
	})
```

```ts
app.OnBeforeServe().Add(e=>{
   e.Router.AddRoute(newRoute({
    Method: http.MethodGet,
    Path: `/api/js`,
    Handler: c=>{
      c.String(200, `Hello from JS`
    }),
    Middlewares: [apis.RequireAdminOrUserAuth()]
})
```

That worked, but something in `echo.MiddlewareFunc` was not behaving correctly. I think the conversion of `RequireAdminOrUserAuth` to JS, then converting the return value from Go to JS, then back to Go again, caused a problem. So instead, I did this:

```go
	m:=map[string] func() echo.MiddlewareFunc {
		"RequireAdminOrUserAuth": apis.RequireAdminOrUserAuth,
	}

	vm.Set("newRoute", func(Method string, Path string, Handler echo.HandlerFunc, Middlewares []string ) echo.Route {
		return echo.Route{
            Method: Method,
            Path:  Path,
            Handler: Handler,
            Middlewares: func()[]echo.MiddlewareFunc{
				var ret []echo.MiddlewareFunc
				for _,v := range Middlewares {
					fn:=m[v]
					if fn!=nil {
						ret = append(ret, fn())
					}
				}
			
				return ret
			}(),
        }
	})
```

```ts
app.OnBeforeServe().Add((e: any) => {
  e.Router.AddRoute(
    newRoute(
      http.MethodGet,
      `/api/js`,
      (c: any) => {
        return c.String(200, 'Hello from JS!')
      },
      ['RequireAdminOrUserAuth']
    )
  )
})
```

Does this seem like the right approach? It feels like a lot of boilerplate considering how many other methods I need to implement. If you have any thoughts on how to improve upon this, I would love to hear it. Thank you!

**Comment by benallfree** [comment]
@ganigeorgiev ;)

**Comment by dop251** [comment]
> 
> But that failed on `AddRoute` because it couldn't convert the JS object back into an `echo.Route` struct. So then I added a constructor function:
>
The problem is not that it can't convert to `echo.Route` (that would work fine), but that the required type is `echo.Routable` which is an interface. There is no way to figure out which concrete type to convert the object to, so you need some sort of a helper function. However, this function can be simply:

```go
	vm.Set("newRoute", func(r echo.Route) echo.Routable {
		return r
	})
```

In this case the parameter will be converted to `echo.Route` automatically. It will be less efficient than your implementation though.

> 
> That worked, but something in `echo.MiddlewareFunc` was not behaving correctly. I think the conversion of `RequireAdminOrUserAuth` to JS, then converting the return value from Go to JS, then back to Go again, caused a problem.

This is more involved. The problem is indeed to do with the fact that the function is converted into JS and back but what goes wrong here is the error propagation. Currently, if a Go function has a single return value of type `error`, an exception within the function is not converted into that error, for that the function must have at least 2 return values. I described the reason for it in #367, but I believe it is a valid case for changing this behaviour. 

In summary, I'm going to do 2 things:

1. Avoid double wrapping, so that `ToValue(func).Export()` returns the original function (as well as `ExportTo()` into an appropriate type).
2. Convert the exception into the `error` return value even if it's a single return value. Also, unwrap `GoError` in such cases so that the original error is returned. In short this would make your case work even if the JS function was not exported into a matching type (or if 1. was not implemented).

**Comment by benallfree** [comment]
Thank you for the detailed answers!

> There is no way to figure out which concrete type to convert the object to, so you need some sort of a helper function.

If goja knows the argument must match the `echo.Routable` interface, couldn't it create the helper function to map the JS object into the struct and throw a runtime error if the object was not the correct type?

Thanks again for looking into this. The ideal outcome would be to simply expose the `app` service to goja and then write convenient wrappers and helpers all in JS.

**Comment by dop251** [comment]
> > There is no way to figure out which concrete type to convert the object to, so you need some sort of a helper function.
> 
> If goja knows the argument must match the `echo.Routable` interface, couldn't it create the helper function to map the JS object into the struct 

Into what struct? How does it know it's `echo.Route` you're after, not some other implementation of `echo.Routable`? It can't even know `echo.Route` exists as a type, there is no runtime package introspection in Go.

**Comment by dop251** [comment]
> I see now. Is it correct to say that any JS object passed into a Go function must be cast using a Go helper function?
> 

No, this is only true for interface types. However there may be other cases where a struct requires initialisation and 
 cannot be simply created as a literal (e.g. `goja.Runtime`) . 

In your case it should work after the fixes I've just pushed, give it a try and let me know if there's any issues.

**Comment by benallfree** [comment]
@dop251 This seems to have worked really well, thank you for the fast fix. 

I think with this change, it just comes down writing helper methods to cast structs to the proper interfaces where needed.
